docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,101 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
+
+The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
```

Comment:
How is nodeKey computed, I remember it's sth like "version+path"?

Reply:
no, it is a just sequenced integer. 
for example, we assign the `nodeKey` of the new node as `tree.nonce + 1` every time when create node

Reply:
It seems better to use "version/seq", where seq is only unique inside the version.

Reply:
here, the `nodeKey` is unique globally.

Reply:
"version/seq" is globally unique as well, seq itself is local unique with the version. For each version, the nodes are written in a batch anyway, we only need to maintain the seq in memory.

Reply:
we keep `version => (root node key = first nonce, last nonce)`, and ensure the nodeKey is unique globally

Reply:
I know, I was just suggesting to use "version/local-seq" as the nodeKey, so we don't need to store this nonce mapping at all.

Reply:
I know there are some advantages of `version|seq`, I am just worried it means we need to update `leftNodeKey` and `rightNodeKey` following this update and it leads to extra encode/decode executions and finally requires more storage to store the node data

Reply:
If we constrain both version and seq to be 32bits, we can squeeze them into 8bytes :)

Reply:
Another thing is "version|seq" can utilize dupsort feature of some db backends.

https://lmdb.readthedocs.io/en/release/

Reply:
> I know there are some advantages of version|seq, I am just worried it means we need to update leftNodeKey and rightNodeKey following this update and it leads to extra encode/decode executions and finally requires more storage to store the node data

Why do we need to update the left and right node keys?

Reply:
here, `leftNodeKey` and `rightNodeKey` refer to the children node key in the db. If we update the node key as `version|seq`, we should also update left and right node keys, right?

Reply:
Meaning that every time we have a new version we update all referenced node keys to refer to that version rather than the original version where they were created? That seems really complex. I feel like I'm maybe not understanding something really basic here

Reply:
@yihuang , I agree with you. If we keep the local nonce as `int32`, there is not much increase in storage size because we can remove the `version` from node body writes.

Reply:
There are several places storage size would change, assuming version is `int64`:
- node key, 8bytes vs 8+4bytes, but with prefix compression, I think the real difference is minimal.
- node body, we can remove the version field, save a `varint(version)`.
- `leftNodeKey` and `rightNodeKey`, `varint(global nonce)` vs `varint(version << 32 + local nonce)`, there'll be a few more bytes here.